### DISCUSSION TITLE
Concerns about plugin security and other things

### DISCUSSION THREAD

**Discussion by Kixunil** [btcpayserver/btcpayserver#3414] [Plugins & integrations]
Firstly, I think this is not a vulnerability serious enough to embargo because it is only exploitable by GitHub or compromised BTCPay developers or other theoretical circumstances and can be avoided mostly by just not using plugins.

I noticed plugins have these security issues:

* They are not signed. GitHub, if compromised, could add backdoored plugins into the repository that a victim could download. It's also possible to do this stealthily (targeted attacks). Similarly, a compromised developer could publish backdoored plugins although not stealthily.
* They run within BTCPayServer process without any sandboxing. It's a shame because sandboxing can significantly improve security. E.g. a compromised plugin could only do limited harm (and conversely some plugin could have higher permissions and BTCPay would have limited access to it). Another downside of current architecture is a crashing plugin brings down whole BTCPayServer as happened in #3407 
* If a user could trick BTCPay into uploading files into plugin directory then he could compromise BTCPay by uploading a malicious plugin. PHP apps are riddled with such bugs and current plugin system kinda reintroduces them.

Other issues:

* LNbank is unusable for users who use readonly+invoice macaroon (CADR and nix-bitcoin by default but there are others I know of). This is somewhat related to sandboxing.
* The user has to restart BTCPay this is inconvenient and can be hard for non-technical users
* (minor) Whitepaper plugin is super-big: 10M compared to 200kB Debian package in CADR that does exactly same thing
* Duplication of functionality: CADR users have two ways of installing whitepaper which could be confusing.
* Plugins may want to interact with things outside BTCPay (system dependencies) and this is not encoded. I plan to write a more detailed document about this topic eventually.
* Alternative sources of plugins still depend on GitHub. It looks like `git clone --depth 1` could be used instead which would also make local sources possible (but being able to just change plugin directory is better)

For me personally these things toghether would resolve all but the sandboxing issue since CADR already takes care of these:

* Make it possible for vendor to disable plugin **store** in **configuration file**
* Make it possible for vendors to change the plugins directory in **configuration file**
* Optional: allow vendor to display custom message in place of plugin store (e.g. a link to their own)

However I strongly recommend to at least sign the plugins for non-CADR users. Also would really, really like to see some sandboxing being implemented.

**Comment by dennisreimann** [comment]
These are good and valid points, getting this right is not trivial for sure -- thanks a lot for this valuable input!

To start small, I began providing SHA256SUMS for LNbank releases (see [v1.1.0](https://github.com/dennisreimann/btcpayserver/releases/tag/BTCPayServer.LNbank%2Fv1.1.0)) and added the files to the [plugin PR](https://github.com/btcpayserver/btcpayserver-plugins/pull/14/files). That gives us something to start with on that front.

**Reply by Kixunil** [reply]
Cool, that's a good first step, now it just needs to be enforced in the BTCPayServer download code.

**Comment by Kixunil** [comment]
> I plan to write a more detailed document about this topic eventually.

I finally did: https://github.com/Kixunil/programming_writings/blob/master/packaging_is_not_dead.md
